core/lib/libtomcrypt/src/modes/ctr/ctr_encrypt.c

```diff
@@ -137,25 +144,29 @@ int ctr_encrypt(const unsigned char *pt, unsigned char *ct, unsigned long len, s
    }
 #endif
 
-   /* handle acceleration only if pad is empty, accelerator is present and length is >= a block size */
-   if (cipher_descriptor[ctr->cipher]->accel_ctr_encrypt != NULL && len >= (unsigned long)ctr->blocklen) {
-     if (ctr->padlen < ctr->blocklen) {
-       fr = ctr->blocklen - ctr->padlen;
-       if ((err = ctr_encrypt_sub(pt, ct, fr, ctr)) != CRYPT_OK) {
-	 return err;
-       }
-       pt += fr;
-       ct += fr;
-       len -= fr;
-     }
-
-     if (len >= (unsigned long)ctr->blocklen) {
+   if (cipher_descriptor[ctr->cipher]->accel_ctr_encrypt != NULL ) {
+     /* handle acceleration only if not in the middle of a block, accelerator is present and length is >= a block size */
+     if ((ctr->padlen == 0 || ctr->padlen == ctr->blocklen) && len >= (unsigned long)ctr->blocklen) {
        if ((err = cipher_descriptor[ctr->cipher]->accel_ctr_encrypt(pt, ct, len/ctr->blocklen, ctr->ctr, ctr->mode, &ctr->key)) != CRYPT_OK) {
-       return err;
+         return err;
        }
        pt += (len / ctr->blocklen) * ctr->blocklen;
        ct += (len / ctr->blocklen) * ctr->blocklen;
        len %= ctr->blocklen;
+       /* counter was changed by accelerator so mark pad empty (will need updating in ctr_encrypt_sub()) */
+       ctr->padlen = ctr->blocklen;
+     }
+
+     /* try to re-synchronize on a block boundary for maximum use of acceleration */
+     incr = ctr->blocklen - ctr->padlen;
+     if (len >= incr + (unsigned long)ctr->blocklen) {
+       if ((err = ctr_encrypt_sub(pt, ct, incr, ctr)) != CRYPT_OK) {
+         return err;
+       }
+       pt += incr;
+       ct += incr;
+       len -= incr;
+       return ctr_encrypt(pt, ct, len, ctr);
```

Comment:
How deep recursion is the worst case?

Reply:
Depth wil not exceed 1.

Indeed, at this point `ctr->padlen == ctr->blocklen`, because `ctr_encrypt_sub()` has just processed `incr` bytes so `ctr->padlen` has been increased by the same amount.
Then two possibilities:
- Either `len >= ctr->blocklen`, in this case the acceleration condition (line 149) will be true, so as many blocks as possible will be processed by the accelerated function, leaving less than (strictly) `ctr->blocklen` bytes to be processed (because of `len %= ctr->blocklen` line 155). Moreover the `ctr->padlen == ctr->blocklen` will still hold (because of line 157). Therefore, the condition at line 162 will not trigger a second time
- Or `len < ctr->blocklen`. In this case since `ctr->padlen == ctr->blocklen` we have `incr == 0`, hence the condition line 162 is also false.
In both cases recursion will end.

I could have called `accel_ctr_encrypt()` and `ctr_encrypt_sub()` instead recursing, but that would have been some significant code duplication.